### PR TITLE
New proc do_refill allows you to tactical reload pill bottles and flare pouches

### PR DESCRIPTION
**Pull request by Xander3359**
## About The Pull Request
You can now refill pill bottles from packets.
## Why It's Good For The Game
Cuts down on some carpal tunnel by allowing you to click a pill bottle with a packet to refill the bottle from the packet.
## Changelog
:cl:
add: You can now reload pill bottles from pill packets.
code: do_refill is on object level, so now it's easy to add tactical reload functionality to containers using refill_types and refill_sound.
/:cl:

### CODE REVIEW

**Comment by ilikesaying** [comment]
basedbasedbasedbasedbased

**Comment by Losenis** [comment]
image

**Comment by FosterGrime-Heymynameslouie** [comment]
TACTICAL MEDICINE

**Comment by tanker4390** [comment]
It may be a poor idea to continue copypasting this proc on high-level objects; perhaps it can instead be broken up into vars and placed lower on the tree? for example
```
.../storage/attackby(I, user)
      if (istype(I, refill_type)
            return do_refill(I, user)
      ...

.../storage/proc/do_refill(I, user)
      this()

.../storage/xeno_holder/rouny_holder/attackby(I, user) //only holds rounys
      if (istype(I, .../rouny_holder))
            var/.../rouny_holder/RH = I
            if (RH.closed)
                  return ..()
            return do_refill(RH, user)
      return ..()

.../storage/xeno_holder/do_refill(I, user)
      for (things in I)
            special_code_for_transferring_xenos(src) //overridden e.g. for automatically creating handfuls
```

**Comment by Araulius** [comment]
it's almost like removing standard pill bottles from being available to marines did nothing during the med rework. shocker.

**Comment by ivanmixo** [comment]
Shhhh dont tell them

**Comment by KerbalCitizen** [comment]
Currently the best method to fill pill bottles with packets is to empty the packet on the floor, then do it again with another packet, and click on the pile of pills with pill bottle. This is really just a QoL PR.

**Comment by Al-1ce** [comment]
Could this be done vice-versa?

**Comment by Xander3359** [comment]
> Could this be done vice-versa?

You cannot refill packets.

**Comment by Autisem** [comment]
could you make it so you can restock pill packets back into the vendor? watching a pile of empty pill packets grow in medbay is kinda jaring.

Other then that, holy fuck finaly

File: code/game/objects/items/storage/firstaid.dm [hunk 1]
```diff
@@ -275,6 +275,33 @@
 	/// Short description in overlay
 	var/description_overlay = ""
 
+/obj/item/storage/pill_bottle/attackby(obj/item/I, mob/user, params)
```
Comment (Lumipharon) [inline code/game/objects/items/storage/firstaid.dm:278]: This should be calling to parent first.

**Review by Lumipharon — CHANGES_REQUESTED** [review]

File: code/game/objects/items/storage/firstaid.dm (hunk 1 above)
Comment (Xander3359) [reply code/game/objects/items/storage/firstaid.dm:278]: done

**Comment by Lumipharon** [comment]
> It may be a poor idea to continue copypasting this proc on high-level objects; perhaps it can instead be broken up into vars and placed lower on the tree? for example
> 
> ```
> .../storage/attackby(I, user)
>       if (istype(I, refill_type)
>             return do_refill(I, user)
>       ...
> 
> .../storage/proc/do_refill(I, user)
>       this()
> 
> .../storage/xeno_holder/rouny_holder/attackby(I, user) //only holds rounys
>       if (istype(I, .../rouny_holder))
>             var/.../rouny_holder/RH = I
>             if (RH.closed)
>                   return ..()
>             return do_refill(RH, user)
>       return ..()
> 
> .../storage/xeno_holder/do_refill(I, user)
>       for (things in I)
>             special_code_for_transferring_xenos(src) //overridden e.g. for automatically creating handfuls
> ```

This is a good point. I didn't realise this is just copypasta from flare refilling, so a storage level proc like this example would be better.

File: code/game/objects/items/storage/storage.dm [hunk 2]
```diff
@@ -519,11 +524,37 @@
 /obj/item/storage/attackby(obj/item/I, mob/user, params)
 	. = ..()
 
+	if(length(refill_types)) //Can anything be emptied into this object
+		for(var/typepath in refill_types)
+			if(istype(I, typepath)) //Is I the right type to refill this object
+				return do_refill(I, user)
```
Comment (Lumipharon) [inline code/game/objects/items/storage/storage.dm:530]:
```suggestion
	if(length(refill_types))
		for(var/typepath in refill_types)
			if(istype(I, typepath))
				return do_refill(I, user)
```

File: code/game/objects/items/storage/storage.dm [hunk 3]
```diff
@@ -519,11 +524,37 @@
...
+		return
+
+	user.balloon_alert(user, "Refilling.")
+
+	if(!do_after(user, 15, TRUE, src, BUSY_ICON_GENERIC))
+		return
+
+	playsound(user.loc, refill_sound, 15, 1, 6)
+	for(var/obj/item/IM in refiller)
+		if(length(contents) >= max_storage_space)
```
Comment (Lumipharon) [inline code/game/objects/items/storage/storage.dm:553]: max storage space is the volume the object can hold, which is not necessarily the number of slots. i.e. a satchel has a max of 15, which translates to 5 normal sized object, or 15 injectors.

**Review by Lumipharon — CHANGES_REQUESTED** [review]

**Review by Lumipharon — APPROVED** [review]

**Pull request merged**